mempool/mempool.go

```diff
@@ -1,26 +1,86 @@
 package mempool
 
 import (
-	"bytes"
-	"container/list"
-	"crypto/sha256"
 	"fmt"
-	"sync"
-	"sync/atomic"
-	"time"
-
-	"github.com/pkg/errors"
 
 	abci "github.com/tendermint/tendermint/abci/types"
-	cfg "github.com/tendermint/tendermint/config"
-	auto "github.com/tendermint/tendermint/libs/autofile"
 	"github.com/tendermint/tendermint/libs/clist"
-	cmn "github.com/tendermint/tendermint/libs/common"
-	"github.com/tendermint/tendermint/libs/log"
-	"github.com/tendermint/tendermint/proxy"
 	"github.com/tendermint/tendermint/types"
 )
 
+// Mempool defines the mempool interface.
+//
+// Updates to the mempool need to be synchronized with committing a block so
+// apps can reset their transient state on Commit.
+type Mempool interface {
+	// CheckTx executes a new transaction against the application to determine
+	// its validity and whether it should be added to the mempool.
+	CheckTx(tx types.Tx, callback func(*abci.Response)) error
+
+	// CheckTxWithInfo performs the same operation as CheckTx, but with extra
+	// meta data about the tx.
+	// Currently this metadata is the peer who sent it, used to prevent the tx
+	// from being gossiped back to them.
+	CheckTxWithInfo(tx types.Tx, callback func(*abci.Response), txInfo TxInfo) error
+
+	// ReapMaxBytesMaxGas reaps transactions from the mempool up to maxBytes
+	// bytes total with the condition that the total gasWanted must be less than
+	// maxGas.
+	// If both maxes are negative, there is no cap on the size of all returned
+	// transactions (~ all available transactions).
+	ReapMaxBytesMaxGas(maxBytes, maxGas int64) types.Txs
+
+	// ReapMaxTxs reaps up to max transactions from the mempool.
+	// If max is negative, there is no cap on the size of all returned
+	// transactions (~ all available transactions).
+	ReapMaxTxs(max int) types.Txs
+
+	// TxsWaitChan returns a channel to wait on transactions. It will be closed
+	// once the mempool is not empty (ie. the internal `mem.txs` has at least one
+	// element)
+	TxsWaitChan() <-chan struct{}
+
+	// TxsFront returns the first transaction in the ordered list for peer
+	// goroutines to call .NextWait() on.
+	// FIXME: leaking implementation details!
+	TxsFront() *clist.CElement
```

Comment:
can't it just return *mempoolTx?

Reply:
or is that the same problem?

Reply:
No, because later in the broadcastTxRoutine loop we're calling `next.WaitNextChan` followed by `next.Next()`. I don't think we want either of these methods in the MempoolTx interface.

What we want - **a simple interface for Reactor to subscribe/get txs from the mempool, so it can broadcast them to peers**.

There are a couple of ways to go about it:

### 1

```go
type Iterator interface {
  Next() bool
  Tx() MempoolTx
```

### 2

```go
TxsChan() <-chan MempoolTx
```
which is used to distribute txs to peers. Exposing a single channel should allow us to remove `TxsWaitChan() <-chan struct{}` method.

Reply:
On the second thought, maybe it's okay for MempoolReactor to require the CList implementation for now. It can be a follow-up PR.

Reply:
https://github.com/tendermint/tendermint/issues/3589